[CSBindings] Split type variable binding inference into phases

xedin opened the pull request:
Refactor `getPotentialBindings` and extract most of the logic to be performed in phased fashion:

Phases include (in order):

- Infer direct bindings (based on all of the constraints associated with the given type variable);
  - Record protocol requirements (both `ConformsTo` and `LiteralConformsTo` constraints)
  - Record `Defaultable` constraints
- Infer transitive bindings based on subtype/conversion relationships;
  - This also infers both protocol requirements and defaults
- Infer default types (if there are no bindings which already cover them);
- Introduce "holes" if necessary.

Resolves: rdar://problem/56212087

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSBindings] Record conformance requirements associated with type variable

Include direct conformnace requirements as well as transitive ones.

xedin committed:
[ConstraintSystem] NFC: Associate transitive binding inference with `PotentialBindings`

xedin committed:
[CSBindings] Add `finalize` method to `PotentialBindings`

This is used to infer transitive information and adjust existing
bindings before solver can consider them.

xedin committed:
[CSBindings] Infer transitive protocol requirements through all conversions

xedin committed:
[CSBindings] Adjust optionality of the compute bindings as a final step

If type variable is expected to conform to `ExpressibleByNilLiteral`
adjust optionality of the inferred bindings only after all of the
bindings have been collected otherwise transitive supertype bindings
are going to stay non-optional which is incorrect.

xedin committed:
[CSBindings] Record holes only after all binding inference steps are done

If there are no direct, transitive or default bindings only then
consider binding the given type variable to a "hole".

xedin committed:
[CSBindings] Infer transitive defaults and add them at finalization

xedin committed:
[CSBindings] Determine literal coverage and default types during finalization

Based on collected direct and transitive information about protocol
requirements let's determine literal protocol coverage by existing
bindings as well as any default types which have to be introduced
to the set as part of finalization phase.

xedin committed:
[CSBindings] Perform more checking before accepting default types

Some literal protocols default to unbound types, we need to make
sure that it's okay to include them to avoid duplicate solutions.

xedin committed:
[CSBindings] Let's always consider `Any` to be a last resort binding

If `Any` has been inferred as a non-default binding (e.g.
if parameter has `Any` type) let's give it the lowest
priority and attempt after any other bindings (including
literal ones) because it's better to infer some concrete
type and then erase it to `Any` if required by the
constraint system.

xedin committed:
[ConstraintSystem] Add a new endpoint to infer bindings for a given type variable

Since bindings now require finalization we need a new endpoint
which perform all of the required actions before returning complete
`PotentialBindings` object when they are requested for a particular
type variable without any other context.

xedin committed:
[CSBindings] Coalesce integer/float literal protocols to produce a single default

Let's try to coalesce integer and floating point literal protocols
if they appear together because the only possible default type that
could satisfy both requirements is `Double`.

xedin committed:
[Diagnostics] Add a tailored fix when ternary is associated with return

If it has been detected that there is an attempt to return a value
from a function without result type, let's diagnose that specifically
in cases when ternary operator is involved.

xedin committed:
[TypeChecker] NFC: Adjust test-cases improved by changes in binding inference

xedin committed:
[TypeChecker] NFC: Add a test-case for fixed rdar://problem/56212087

xedin commented on test/SILGen/function_conversion.swift:
```diff
@@ -663,14 +663,14 @@ struct FunctionConversionParameterSubstToOrigReabstractionTest {
+// CHECK-NEXT: store [[DOWNCAST_RESULT]] to [init] [[ADDR]] : $*Array<(AnyHashable, Any)>
+// CHECK-NEXT: // function_ref Dictionary.init<A>(uniqueKeysWithValues:)
+// CHECK-NEXT: [[DICT_INIT:%.*]] = function_ref @$sSD20uniqueKeysWithValuesSDyxq_Gqd__n_tcSTRd__x_q_t7ElementRtd__lufC
+// CHECK-NEXT: apply [[DICT_INIT]]<AnyHashable, Any, [(AnyHashable, Any)]>([[ADDR]], [[TYPE:%.*]])
```
We used to add a thunk because one of the bindings would incorrectly have `SubtypeOfExistentialType` set to `true` when `Bindings` where empty.

xedin commented:
One more think I forgot to do is change it so empty bindings are not included, I'm going to adjust that now.

xedin commented:
I remembered why I can't do that and everything has to be included - even though bindings might not have anything it's possible to transitively infer protocol requirements and defaults and pass them along the subtype chain.

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
This is promising, I'm going to try full suite next.

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test compiler performance

xedin commented:
@swift-ci please test source compatibility debug

xedin commented:
@swift-ci please test compiler performance

xedin commented:
@swift-ci please test compiler performance

xedin commented:
@swift-ci please test compiler performance

xedin committed:
[CSBindings] Limit optionality hack only to direct protocol requirements

In situations like:

$T0 subtype $T1
$T1 literal conforms to ExpressibleByArrayLiteral
$T0 conv [<Type>]?

We have to ensure that $T0 gets to maintain its optionality
because $T1 then could strip optionality later based on a
typing rule where `T <: T?`

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

hborla commented on lib/Sema/CSBindings.cpp:
```diff
@@ -75,13 +83,284 @@ void ConstraintSystem::inferTransitiveSupertypeBindings(
+    });
+
+    if (anyType != Bindings.end()) {
+      std::rotate(Bindings.begin(), anyType + 1, Bindings.end());
```
If the order of `Bindings` is important, why is a rotation okay here? Why isn't this moving _only_ the `Any` binding to the end?

xedin replied:
`std::rotate` is only moving `Any` to the end, it should replace its current position with next in the iterator and move it to after the .end(). Maybe I misunderstood that API though...

xedin replied:
Ah I think I understand what is the problem, here, I'd need some other way to only move one element...

xedin replied:
Actually, it looks like what I need to do is `std::rotate(anyType, anyType + 1, Bindings.end());`

hborla commented on lib/Sema/CSBindings.cpp:
```diff
@@ -75,13 +83,284 @@ void ConstraintSystem::inferTransitiveSupertypeBindings(
+
+    // We need to figure out whether this is a direct conformance
+    // requirement or inferred transitive one to identify binding
+    // kind correctly.
```
I don't fully understand this - could you elaborate a little on why this is the case?

xedin replied:
Sure! If the conformance requirement is inferred transitively solver should be allowed to try supertypes of default type, if it’s directly associated with the current type variable solver is allowed to try subtypes of default. Does that make more sense?

xedin replied:
So if we have `$T0 <: $T1` and `$T0 conforms to ExpressibleByIntegerLiteral` for example, when inferring a default for `$T1` from the conformance requirement on `$T0` solver should be allowed to attempt supertypes just like if it was `Int <: T1`.

hborla replied:
Ohhh okay, I was thinking about transitive bindings the other way around (e.g. the conformance is on the supertype instead of the subtype). This makes sense. I realize now that the documentation for `inferTransitiveBindings ` would have cleared up my confusion haha. Thank you!

xedin committed:
[CSBindings] Fix `std::rotate` to move only `Any`

It used to move everything up to `Any`, now it's going to move
only `Any` itself to the end of the list.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
Alright, I'm going to merge this and fix any problems as they come up.
